src/Microsoft.Health.Core/Features/Events/EventGridEventGenerator.cs

```diff
@@ -34,6 +37,28 @@ public EventGridEventGenerator(Uri endpoint, string key)
             _client = new EventGridPublisherClient(endpoint, new AzureKeyCredential(key));
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEventGenerator{T}"/> class.
+        /// EventGridEventGenerator constructor.
+        /// </summary>
+        /// <param name="endpoint">Uri.</param>
+        /// <param name="publisherCertificate">publisher certificate.</param>
+        public EventGridEventGenerator(Uri endpoint, X509Certificate publisherCertificate)
+        {
+#pragma warning disable CA2000 // Dispose objects before losing scope
```

Comment:
This would leak resources. Implement IDisposable for EventGridEventGenerator ....keep the handler as member and then dispose properly. same for the next.